[Tree]미션 3 체스판 초기화

choitree opened the pull request:
미션1에 브랜치 관리를 잘못하여, 미션2 내용이 포함되었습니다.
그래서 미션1 이후에 미션3으로 pr 요청드립니다.
romote와 rebase 작업으로 브랜치가 생성되지 않은 상태로 pr 요청드리며
미션4부터는 브랜치를 별도로 생성해서 진행하도록 하겠습니다.

choitree committed:
feat : Do Mission2

Create Board and BoardTest

choitree committed:
feat : Do Mission2

Create Board and BoardTest

choitree committed:
feat : Do Mission2

Create Board and BoardTest

choitree committed:
feat : Do Mission3

Initialize Chess Board
Print Chess Board using StringBuilder
Make MainApp and input message what to do

choitree committed:
feat : Do Mission3

Initialize Chess Board
Print Chess Board using StringBuilder
Make MainApp and input message what to do

choitree committed:
fix : Fix mission1

다시 리뷰받은 내용 반영해서 커밋합니다.

choitree committed:
.

choitree committed:
.

choitree committed:
rebase mission3

choitree committed:
Merge branch 'choitree' of https://github.com/choitree/java-chess into choitree

choitree committed:
fix : mission3
브랜치 romote를 하지 않아, romote 및
rebase 작업하였습니다.

choitree committed:
Merge branch 'choitree' into choitree

kyupid commented on src/main/java/net/tree/Main.java:
```diff
@@ -0,0 +1,35 @@
+public class Main {
+    public static void main(String[] args) {
+
+        boolean flag = true;
```
`boolean flag` 를 따로 선언해주신 이유가 있으신가요?
제생각에는 flag빼고 while 문에 true로해도 똑같아보여서요

choitree replied:
exit에서 나갈라고 하면 false로 바꿔야해서 flag를 선언했어요~~
근데 생각해보니 어차피 종료하는 코드가 있으니까 선언할 필요가 없긴하네요!!!

kyupid commented on src/main/java/net/tree/Main.java:
```diff
@@ -0,0 +1,35 @@
+
+        boolean flag = true;
+        while (flag) {
+            Scanner sc = new Scanner(System.in);
```
Scanner close하는 부분이 없는거같아요
제가 리뷰받은걸 참고해서 말씀드리면[ try-with-resource](https://codechacha.com/ko/java-try-with-resources/)를 이용하면 쉽게 관리할수있다고합니다

choitree replied:
Scanner close를 안해도 자동으로 닫힌다고 예전에 쿠퍼가 알려줘서
따로 안닫았는데 참고해서 반영해볼게요!! 고마워요~~~👍

kyupid replied:
엇 Scanner close 안해도 자동으로 닫힌다구요?? 나중에 저도 다시확인해보겠습니다

honux77 replied:
슬랙에도 있지만 습관적으로 닫아주는게 좋습니다. Scanner(System.in) 이면 사실 안 닫아줘도 큰일은 안 납니다.

kyupid commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -8,7 +8,8 @@
     List<Pawn> pawnList = new ArrayList<>();
-
+    final int BOARD_SIZE = 8;
+    String blankDot = "........\n";
```
제 생각에 `blankDot`은 클래스 내에서만 사용해도 될거같은데 접근제어자 default로 하신 이유가 있으신가요??

kyupid commented on src/main/java/net/tree/pieces/Pawn.java:
```diff
@@ -1,22 +1,36 @@
+    public char getRepresentation() {
+        return representation;
+    }
+}
```
36라인에 보시면 No newline at end of file 이라고 나오는데요
.
항상 파일 마지막에 개행해줘야한다고 합니다
아이작이 블로그에 이에관해서 [포스트한 글 링크](https://isaac56.github.io/intellij/2021/02/09/setting_newLine_at_EOF.html) 올릴게요

kyupid commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -30,10 +30,37 @@ public void create() {
+
+    @DisplayName("보드 초기화 및 그리기")
+    @Test
+    public void initializeBoardAndPrint() {
+        Board board = new Board();
+        board.initialize();
+        board.print();
+    }
+
+    @DisplayName("보드 초기화하기")
+    @Test
+    public void initialize() {
+        Board board = new Board();
+        board.initialize();
+        assertEquals("pppppppp", board.getPawnsResult(Pawn.WHITE_REPRESENTATION));
+        assertEquals("PPPPPPPP", board.getPawnsResult(Pawn.BLACK_REPRESENTATION));
+    }
+
     @DisplayName("pawnList 확인해보기 ")
     @Test
     public void checkPawnList() {
         Board board = new Board();
```
`Board board = new Board();` 의 중복을 `@BeforeEach` 를 통해서 줄일수있습니다

kyupid commented on src/test/java/net/tree/pieces/PawnTest.java:
```diff
@@ -1,30 +1,39 @@
+
     }
+
 }
```
여기도 No newline at end of file 이라고 뜨네용

kyupid reviewed (commented):
안녕하세요 트리 ㅎㅎ 오랜만이죠?? 제가 아는 한도내에서 열심히 코멘트해봤습니다 도움되시길!

choitree committed:
kyu가 리뷰해준 내용 반영하였습니다.

ksundong commented on build.gradle:
```diff
@@ -9,9 +9,8 @@ repositories {
-    //testCompile 'junit:junit:4.12'
-    testCompile 'org.junit.jupiter:junit-jupiter:5.4.2'
+
 }
```
https://github.com/codesquad-members-2021/java-chess/pull/9#discussion_r572027344
이 코멘트 참고하셔서 IntelliJ에 설정해주세요!

choitree replied:
java 파일 뿐 아니라 모든 파일에 다 개행이 필요한거였군요!! 반영했습니다 감사합니다

honux77 replied:
//testCompile 'junit:junit:4.12' 이 부분 주석도 필요없으니 제거하는게 좋겠어요.

choitree replied:
네네 삭제했습니다~~ junit assertj 와 의존성 설정 시 컴파일/임플리먼트에 대한 개념이 전혀 없는 상태였는데
이거저거 바꿔보면서 어떤 기능인지 느낌인지 대충  와닿은 것 같습니다(아직 갈길이 구만리)

ksundong commented on src/main/java/net/tree/Main.java:
```diff
@@ -0,0 +1,31 @@
+                switch (inputMessage) {
+                    case  "start" :
+                        System.out.println("====게임을 시작합니다.====\n");
+                        Board board = new Board();
```
start 할 때 마다 새로운 보드가 생성되어야 할까요?

choitree replied:
혹시 말씀해주신 버그가 이거인가요?~~ 아님 이것 포함 플러스 알파인가요?ㅠㅠ
아까 메인클래스로 테스트를 해보다가 저도 이러면 계속 보드가 생기는건데...하고 아차 싶었습니다

choitree replied:
start할때마다 pawn이 늘어나서, 생성과 initalize를 while문 밖으로 뺏습니다

ksundong replied:
이 부분도 버그라고 볼 수 있지만, black을 생성시 white가 생성되는 부분을 의미한게 더 컸습니다~

ksundong commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -8,7 +8,8 @@
 public class Board {
 
     List<Pawn> pawnList = new ArrayList<>();
-
+    final int BOARD_SIZE = 8;
```
접근제어자에 신경써주세요~

ksundong commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -8,7 +8,8 @@
     List<Pawn> pawnList = new ArrayList<>();
-
+    final int BOARD_SIZE = 8;
+    private String blankDot = "........\n";
     public void add(Pawn pawn) {
```
둘 사이에 빈 라인을 추가해주세요~

ksundong commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -20,7 +21,44 @@ public int checkPawnSize() {
         return pawn;
     }
+
+    public List<Pawn> initialize() {
```
initialize가 pawnList를 반환해야하는 이유가 있을까요?

choitree replied:
반환을 해줘야 pawnList에 담길거라고 생각하고 만들었던 것 같은데, 다시 생각해보니
void로 해도 상관없네요~ 수정하도록 하겠습니다!

ksundong commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -20,7 +21,44 @@ public int checkPawnSize() {
+        StringBuilder sb = new StringBuilder();
+
+        for(int i = 0 ; i < pawnList.size() ; i++) {
+            if(pawnList.get(i).getRepresentation() == representation) {
```
이 조건식은 왜 들어가야했는지 잘 이해하지 못했어요. 설명좀 부탁드릴게요~

choitree replied:
pawn이 아닌 다른 피스도 한 리스트에서 관리하려고 하면서, 그러면 pawn인지 아닌지 확인이 필요할 것으로 판단되어
미리 작성한 조건식인데 현재로서는 불필요한 코드였습니다.
호눅스가 미리 예상하고 작성하지 말라고했는데 저도 모르게 작성했네요 흡...

choitree replied:
아!! 코드를 수정하다가 보니까 왜 이렇게 작성했는지 기억났어요!!!
저는 pawnList 한개에 색상 무관하게 다 집어넣기 때문에 for문으로 16개 size로 도는데
만약에 조건문 없이 append를 한다면 16개의 representation이 생성되요
그래서 representation을 매개변수로 받아서, 그 인자와 리스트에 있는 pawn의 representation이 같으면
append되도록 하려고 if문을 사용했습니다~

ksundong replied:
음 말씀하시면서 꺠달으셨을 수도 있는데, 색상을 분리해서 넣는 방안도 한 번 생각해보세요.
색상을 받아서 같아야만 더하는건 복잡하다고 생각됩니다~

ksundong commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -20,7 +21,44 @@ public int checkPawnSize() {
+        return sb.toString();
+    }
+
+    public String saveStringBuilder() {
```
이 메서드명은 그다지 좋지 못한 네이밍이라고 생각됩니다.
이 메서드가 하는 일을 더 잘 표현하는 이름으로 바꿔봅시다.

ksundong commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -20,7 +21,44 @@ public int checkPawnSize() {
+    }
+
+    public void print() {
+        System.out.println(saveStringBuilder());
```
이 `System.out.println()` 부분은 `Main`에 두는 편이 좋을 것 같습니다.

ksundong commented on src/main/java/net/tree/pieces/Pawn.java:
```diff
@@ -1,22 +1,36 @@
 
 public class Pawn {
+
     String color;
+    char representation;
```
접근 제어자에 신경써주세요!

ksundong commented on src/main/java/net/tree/pieces/Pawn.java:
```diff
@@ -1,22 +1,36 @@
     }
 
-
     public Pawn(String color) {
+        this.color = WHITE_COLOR;
+    }
```
이 생성자는 조금 잘못된 것 같은데요~ ^^

ksundong commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -10,11 +11,17 @@
 
 public class BoardTest {
```
JUnit5에서는 Test class에 `public`을 붙여주지 않는 것이 더 일반적입니다.

ksundong replied:
이 부분은 모든 테스트코드 다 확인하셔서 고쳐주세요!

ksundong commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -10,11 +11,17 @@
 
+    Board board;
+
+    @DisplayName("new Board() 중복코드 줄이기 위해 사용.")
```
상황적 맥락을 설명해주는 것도 좋지만 그냥 안써줘도 좋을 것 같습니다 ㅎㅎ

choitree replied:
DisplayName을 작성하는 것을 통일하기 위해서 작성했는데, test 어노테이션에만 작성하면 될까요??

ksundong replied:
이 테스트가 하는 일을 설명해주는 것이 좋다고 생각합니다.
상황적 맥락은 하는 일보단 덜 중요하거든요~

ksundong commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -30,10 +37,35 @@ public void create() {
+    @DisplayName("보드 초기화 및 그리기")
+    @Test
+    public void initializeBoardAndPrint() {
+        board.initialize();
+        board.print();
```
이 테스트 코드는 무엇을 어떻게 테스트하는 코드인가요?
자동화된 테스트에는 테스트 한 내용이 검증되는 코드가 있어야 합니다.

choitree replied:
이 코드는 보드가 잘 Print되는지 확인을 위해서 작성했던 코드인데
단순히 프린트 여부를 확인하기 위한 메소드는 test 코드로 작성하지 않는 것이 일반적인가요?~

ksundong replied:
네 출력값을 검증할 수도 있긴 하지만, 출력값을 테스트하기 보다는 출력될 값을 테스트하는 편이 테스트하기에도 더 좋고, 책임을 분리하는데에도 도움이됩니다.

ksundong commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -30,10 +37,35 @@ public void create() {
+
+    @DisplayName("보드 초기화하기")
+    @Test
+    public void initialize() {
```
`JUnit5`에서는 테스트 메서드에 `public`을 붙여주지 않는 것이 일반적입니다. 참고해주세요!

ksundong replied:
이 부분 역시 모든 테스트 코드 확인 하시고 고쳐주세요!

ksundong commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -30,10 +37,35 @@ public void create() {
+    @Test
+    public void initialize() {
+        board.initialize();
+        assertEquals("pppppppp", board.getPawnsResult(Pawn.WHITE_REPRESENTATION));
+        assertEquals("PPPPPPPP", board.getPawnsResult(Pawn.BLACK_REPRESENTATION));
```
`assertThat()`과 같은 assertJ의 기능을 사용해주세요.

ksundong commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -30,10 +37,35 @@ public void create() {
+        assertEquals(1, board.checkPawnSize());
+        assertEquals(white, board.findPawn(0));
+
+        Pawn black = new Pawn(Pawn.BLACK_COLOR);
```
이거 정말 검정색 Pawn이 생성될까요?

choitree replied:
아니요 무족권 white.....😧

ksundong commented on src/test/java/net/tree/pieces/PawnTest.java:
```diff
@@ -1,30 +1,39 @@
-    public void create() {
-        verifyPawn(Pawn.WHITE_COLOR);
-        verifyPawn(Pawn.BLACK_COLOR);
+    public void create_기본생성자() {
```
저는 개인적으로 한/영 혼용은 추천드리지 않습니다.

honux77 replied:
맞아요. 차라리 그냥 한글이 좋을 듯

ksundong commented on src/test/java/net/tree/pieces/PawnTest.java:
```diff
@@ -1,30 +1,39 @@
-        verifyPawn(Pawn.WHITE_COLOR);
-        verifyPawn(Pawn.BLACK_COLOR);
+    public void create_기본생성자() {
+        Pawn pawn = new Pawn();
+        Assertions.assertAll(
+                () -> assertEquals(Pawn.WHITE_COLOR, pawn.getColor()),
+                () -> assertEquals(Pawn.WHITE_REPRESENTATION, pawn.getRepresentation())
+        );
     }
 
-    void verifyPawn(final String color) {
-        Pawn pawn = new Pawn(color);
-        assertEquals(color, pawn.getColor());
+    @DisplayName("color, representation 매개변수로 받아서 생성한 pawn")
+    @Test
+    public void create() {
+        verifyPawn(Pawn.WHITE_COLOR, Pawn.WHITE_REPRESENTATION);
+        verifyPawn(Pawn.BLACK_COLOR, Pawn.BLACK_REPRESENTATION);
 
     }
```
생성자가 총 3개인데, 테스트가 있는 것은 2개네요.
주의해줍시다!

choitree replied:
color만 받는 생성자를 삭제하고 그에 맞게 수정하였습니다

ksundong commented on src/test/java/net/tree/pieces/PawnTest.java:
```diff
@@ -1,30 +1,39 @@
+        Pawn pawn = new Pawn(color, representation);
+
+        Assertions.assertAll(
+            () -> assertEquals(color, pawn.getColor()),
+            () -> assertEquals(representation, pawn.getRepresentation())
```
여기도 마찬가지로 assertJ의 API를 사용해봅시다.

ksundong reviewed (changes requested):
안녕하세요! Tree 리뷰어 Dion입니다. 미션 3을 진행해주시느라 고생많으셨습니다! 👍
몇 가지 코멘트 드릴 사항이 있습니다. 참고해주세요.

## PR 작성 관련

- 가급적이면, 커밋 메시지는 학습 과정이니만큼 상세하게 작성해주세요. https://meetup.toast.com/posts/106 이 포스팅을 한 번 읽어봐주세요. 커밋 메시지는 한글로 작성하셔도 무방합니다!
- 그리고 커밋 메시지는 일관적으로 작성해주세요. 템플릿을 만드셨는데, 그 규칙을 준수해주세요. '.'과 같은 커밋 메시지는 절대 사용하지 말아주세요.
- PR에는 상황을 설명해주는 것도 좋지만, 학습 과정에서 겪었던 어려움, 궁금한점을 남겨주시면 리뷰어가 도움을 드릴 수 있습니다.
- 가급적이면 이런 경우에는 그냥 PR을 올리기보다는 새로 브랜치를 작성하시고, 정돈된 PR을 올려주시는 편이 좋습니다. 커밋 기록이 그대로 남기 때문이에요.

## 코드 관련

- 접근 제어자의 사용이 아직 미숙합니다. 신경을 꼭 써주세요.
- 반환 타입이 왜 필요한 것인지 더 생각을 하고 메서드를 작성해주세요.
- 메서드, 변수의 네이밍은 조금 어렵지만, 그래도 읽는 사람을 생각한 네이밍을 시도해주세요.
- 가급적이면 현재 체스 과정에서는 모든 코드를 커버할 수 있도록 테스트를 잘 작성해주세요. 현재, 버그가 있는 부분이 있어서 말씀드립니다.
- 테스트 코드는 무엇을 테스트 할 것인가를 명확하게 해주세요!
- 현재 교육과정에서 assertJ의 사용을 목표로 하고 있기 때문에 assertJ의 사용을 권장드립니다.
- 네이밍 시에 한 영 혼용은 추천드리지 않습니다.

리뷰가 늦어져서 죄송합니다. 미션 진행하느라 고생많으셨어요.
구현에 조금 문제가 있는 부분이 발견되어 변경요청 드렸고, 코멘트 된 부분도 한 번씩 확인해주세요!!
궁금하시거나, 이해가 잘 안가시는 부분이 있으면 코멘트 남겨주시고, 수정이 완료되시면 다시 리뷰요청해주세요!

choitree committed:
Dion이 리뷰해주신 내용 반영하다.

리뷰받은 내용 수정하였다.
1.접근제어자에 신경이 필요하다. private/public 주의
2.다음 미션으로 넘어갈 때, 불필요한 코드 삭제해서 에러 줄이기
3.생성자가 여러개라면, 같은 개수로 test코드 작성 필요
4.assertj 사용!!!
5.JUnit에서는 접근제어자 작성 불필요
6.메소드명은 영문으로 작성
7.return 타입 생각해서 작성
8. 메소드명, 변수명 읽는 사람도 고려해서 상세히 작성

choitree committed:
BoardTest 내 메소드명 수정

질문을 하기 위해 임의로 작성한 메소드명을 수정하지 않고 커밋하여
해당 메소드명을 수정하였습니다.

choitree reviewed (commented):
Dion 상세한 리뷰 정말 감사합니다. 첫 단락에 작성해주신 아예 메모해서 모니터에 부착해뒀습니다!!
코드 작성할 때마다 염두해두고 작성하도록 하겠습니다.
다른 내용은 다 수정했고, 제가 미션이 넘어갈 때마다 이전 코드에 미련이 남아서 
이전 미션에서 요구했던 내용을 삭제하지 않아서 이에 따른 버그나 오류가 많이 있었던 것으로 보여졌습니다.
미션4에서는 해당 미션에 해당되는 내용이 아니라면 미련없이 삭제하고 있습니다.
읽으시느라 고생 많으셨을텐데 죄송합니다ㅠㅠㅠㅠ

제 메소드명이 전반적으로 불친절한 것 같아서, 알려주시지 않은 않은 부분도
메소드명을 수정해보았습니다. 그런데 작성 중 궁금한 점이 있어서 여쭤봅니다.

메소드명은 작성자 이외의 사람이 읽기 좋게 자세하게 작성하는 것인데, 그 리미트를 가늠하기가 어렵습니다.
ex)BoardTest 마지막 메소드같은 경우, 리스트에 pawn이 각각 잘 들어가는지 확인하기 위한 test인데
작성자가 아니면 테스트를 자세히 들여다봐야지 의도롤 파악할 수 있을 거 같아서 작성하다보니
checkPawnListWithWhitePawnAndBlackPawn() 라고 작성하였습니다.(말도안돼...)
오히려 너무 긴 메소드명은 가독성이 떨어지는 것 같은데, 
테스트 코드는 DisplayName 어노테이션을 사용한다고 해도, 
일반 클래스에 메소드명은 어느 정도 길이로 작성해야하는 것이 좋을까요?~~

choitree committed:
Main.java에 초기화 메소드 호출 부분 수정

board 생성과 함께 board 초기화도 start 바깥으로
이동했었는데, 시작할 때마다 초기화가 필요하기 때문에
메소드 호출 위치 다시 start 케이스 내로 이동

ksundong commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -8,8 +8,10 @@
 public class Board {
 
     List<Pawn> pawnList = new ArrayList<>();
+    private final int BOARD_SIZE = 8;
```
이 변수명은 컨벤션을 어기고 있습니다.
어떤 경우에 이러한 형태의 변수를 써야하는지 학습해주세요~

choitree replied:
음 제가 알기로는 상수는 대문자로 작성하고, 단어 사이에 언더스코어를 사용한다고 알고있고
자바 상수 컨벤션을 찾아봐도 유사하게만 나오는데 혹시 어떤 키워드로 학습해야하는지 알려주실 수 있나요??

ksundong replied:
(클래스 변수)는 `static`이라는 키워드를 사용해야겠죠?
유사한게 아니라 일치 하는 형식으로 작성해주시면 되겠습니다!

choitree replied:
아 static이 빠진지 몰랏네요!! 저는 이름에만 집중해가지고 static이라고 생각을 못했네요ㅠㅠ

ksundong commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -8,8 +8,10 @@
 
     List<Pawn> pawnList = new ArrayList<>();
+    private final int BOARD_SIZE = 8;
+    private String blankDot = "........\n";
```
blankDot이라는 이름은 조금 의아할 수 있을 것 같아요.
Line이나 Rank(?) 정도로 바꿔봅시다.

choitree replied:
공백을 별로할까 점으로 할까 고민하다가, 점으로 찍어야지!하면서 의식의 흐름으로 네이밍을 해버렸네요 하하
항상 일단 작성하고 나중에 수정해야지 생각하는데, 막상 수정하는 일은 극히 드물고 코드 리뷰 받으면서 더 실감하게되서
처음에 생각하면서 잘 짓고 만들어야겠다는 생각을 많이 하게됩니다

ksundong commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -8,8 +8,10 @@
+    private String blankDot = "........\n";
 
-    public void add(Pawn pawn) {
+    public void addPawntopawnList(Pawn pawn) {
```
항상 camelCase는 지켜주세요.
`addPawnToPawnList`

choitree replied:
넵넵!

ksundong commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -20,7 +22,40 @@ public int checkPawnSize() {
+        return sb.toString();
+    }
+
+    public String appendBoardWithPawns() {
```
저라면 `getRepresentation` 정도로 해줬을 것 같아요.
메서드 안에서 하는 일을 설명하는게 아니라, 이 메서드의 역할을 설명해주세요.

choitree replied:
pawn을 보이게하는걸 메소드명에 드러내고 싶어서, getPawnsRepresentation로 변경했습니다!

ksundong commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -20,7 +22,40 @@ public int checkPawnSize() {
+    }
+
+    public String appendBoardWithPawns() {
+        StringBuilder sb = new StringBuilder();
```
`sb` 같은 네이밍보다 더 좋은 이름을 고민해봅시다.

choitree replied:
StringBuilder에서 sb라고 사용하는거나, Scanner에서 sc라고 사용하는 것처럼 많이들 사용하는 네이밍은
남들도 그렇게 쓰기도 하고 그리고 길어지면 메소드 내에서 사용 시 오히려 가독성이 떨어진다고 생각해서
간략하게 작성하기 위해서 사용하기도 했습니다

현재는 메소드명에 맞춰서 pawnsRepresentation로 변경했는데
이런 식으로 길더라도 하는 일을 적시해주는 것이 좋은 것인가요?

ksundong replied:
단순히 sb로 네이밍하는 것 보다는 훨씬 좋죠.
또한 학습단계이기 때문에 이상적인 방식으로 작성하는게 좋은 것 같다고도 생각해요.

ksundong commented on src/main/java/net/tree/pieces/Pawn.java:
```diff
@@ -1,22 +1,32 @@
-
-    public Pawn(String color) {
+    public Pawn(String color, char representation) {
         this.color = color;
+        this.representation = representation;
```
자, 이렇게 들어오면, color가 white지만, black representation이 들어올수도 있고
혹은 전혀 관계없는 문자열과 문자가 들어올 수 있습니다.
이는 나중 단계에 포함되어 있지만, 항상 주의해주세요.

choitree replied:
그러면 생성자 내에서 if문으로 validation 을 주어도 괜찮을까요?

```
public Pawn(String color, char representation) {
        if(color.equals(WHITE_COLOR) && representation == WHITE_REPRESENTATION) {
            this.color = color;
            this.representation = representation;
        }
        else if(color.equals(BLACK_COLOR) && representation == BLACK_REPRESENTATION) {
            this.color = color;
            this.representation = representation;
        }
        else {
            throw new IllegalArgumentException();
        }
        
    }
```

이런식으로 validate하는것도 괜찮은 방법인가요? 
이번 미션에서는 코드를 변경하지 않고 다음단계에서 디벨롭해보려고 여기에 작성해서 여쭤봅니다~
근데 저런식으로 작성하면 pieces에서 적용하면 엄청 긴 코드가 되겠네요😳
조금 더 간결하게 만들 방법을 강구해봐야겠습니다

ksundong replied:
복잡해지죠? 다음 단계에서 개선해봅시다.

ksundong commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -1,39 +1,68 @@
-
-        Pawn white = new Pawn(Pawn.WHITE_COLOR);
-        board.add(white);
+    void createWhitePawnAndBlackPawn() {
```
훨씬 좋아졌네요 👍

ksundong commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -1,39 +1,68 @@
-    public void checkPawnList() {
-        Board board = new Board();
-        assertTrue(board.pawnList.add(new Pawn()));
+    void initializeBoardAndAppend() {
```
Append보다 좋은 네이밍을 고민해봐주세요~

choitree replied:
테스트 기능과 함께 DisplayName, 메소드명 수정했습니다

ksundong commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -1,39 +1,68 @@
-        assertTrue(board.pawnList.add(new Pawn()));
+    void initializeBoardAndAppend() {
+        board.initialize();
+        board.appendBoardWithPawns();
```
해당 테스트는 당연히 통과합니다.
그럼 변경사항이 생겼을 때, 실패하는지는 어떻게 확인할 수 있을까요?

choitree replied:
제가 작성한 것은 테스트 기능이 없는 코드였네요! 
assertThat으로 수정했습니다~

ksundong commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -1,39 +1,68 @@
+
+    }
+
+    @DisplayName("whitePawn과 blackPawn 1개씩으로 pawnList 확인해보기 ")
+    @Test
+    void checkPawnList() {
+        Pawn white = new Pawn(Pawn.WHITE_COLOR, Pawn.WHITE_REPRESENTATION);
+        board.addPawntopawnList(white);
+        assertThat(board.checkPawnSize()).isEqualTo(1);
+        assertThat(board.findPawn(0)).isEqualTo(white);
+
+        Pawn black = new Pawn(Pawn.BLACK_COLOR, Pawn.BLACK_REPRESENTATION);
+        board.addPawntopawnList(black);
+        assertThat(board.checkPawnSize()).isEqualTo(2);
+        assertThat(board.findPawn(1)).isEqualTo(black);
+    }
```
위에서 하는 테스트와 차이가 있나요?

choitree replied:
아니요 같은건데 이것도 미션2의 잔재였네요ㅠㅠ 삭제했습니다!

ksundong reviewed (changes requested):
Tree 수정하시느라 고생하셨습니다. 필요없어진 코드는 과감히 지워주세요.
메서드 명은 좋은 방향으로 바뀌었다고 생각합니다.

말하신 대로 테스트 메서드명은 너무 길어도 좋지 않습니다.
길이는 딱히 정해진 것은 없습니다. 나중에 추상화라는 개념을 배우게 될 텐데, 그러면 메서드명의 적절한 길이를 생각해볼 수 있을 것 같아요.

메서드명은 이 클래스, 객체가 하는 일을 대략적으로 알려주는 역할을 합니다.
내부 구현을 메서드명에 드러내지 않고, 하는 일만 나타내도록 해주시면 되겠습니다.
그렇게 설명했을 때, 과다하다면 책임이 여러개는 아닌지 의심해 볼 수 있습니다.
책임까지 분리했는데도 길다면, 그것이 메서드 이름의 적절한 길이라고 짐작할 수 있습니다.

수정하신 내용은 좀 더 학습하시면 좋겠다고 판단하여, 다시 변경요청을 드립니다.
공부하시고, 수정하시느라 고생많으셨습니다~

choitree committed:
Dion의 두번째 리뷰 내용 반영

첫번째 PR에서 부가적인 내용을 리뷰받아 반영하다.
-변수명, 메소드명 수정
-테스트 기능이 없는 코드에 assertj로 추가
-기능에 맞게 메소드명 수정하고, DisplayName 변경
-불필요한 테스트 코드 삭제

choitree reviewed (commented):
리뷰해주신 내용을 다시 반영해보았습니다.
확실히 혼자 코드를 짜는 것보다 리뷰를 해주신 내용을 토대로 학습 키워드를 얻어서 찾아보고 알아가는게 흥미롭네요!
그런데 생성자 관련해서 리뷰해주신 부분은 Dion의 말씀 의도를 제가 제대로 이해한 것인지 모르겠어서
그 부분에 대해서 코드와 함께 다시 여쭤봤습니다~

이번에는 merge되기를 (두구두구두구)

ksundong commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -8,8 +8,10 @@
 public class Board {
 
     List<Pawn> pawnList = new ArrayList<>();
```
접근제어자도 넣어주세요~

ksundong commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -1,39 +1,66 @@
-        assertTrue(board.pawnList.add(new Pawn()));
+    void checkInitializeBoard() {
+        board.initialize();
+        StringBuilder initializedBoard = new StringBuilder();
+        initializedBoard
+                .append("........\n")
+                .append("PPPPPPPP\n")
+                .append("........\n")
+                .append("........\n")
+                .append("........\n")
+                .append("........\n")
+                .append("pppppppp\n")
+                .append("........\n");
```
굳이 StringBuilder를 쓰지 않아도 좋을 것 같아요~
변경 안하셔도 됩니다~

ksundong reviewed (approved):
안녕하세요. Tree! 제가 코멘트 남긴 부분을 수정해주세요.
여기에 해당 변경사항 반영해주시면, Merge하겠습니다.
이번 단계 수행해주시느라 고생하셨습니다!!

honux77 commented:
둘 다 수고 많았어요! 화이팅!

choitree committed:
Dion의 세 번째 리뷰 내용 반영(2번째 PR 마지막 리뷰)

첫번째 PR에서 부가적인 내용을 리뷰받아 반영하다.
-상수 네이밍 컨벤션 수정(static!!)
-List에 접근제어자 수

choitree commented:
엇 모르고 닫아버렸어요ㅠㅠ 수정하고 푸시했습니다~~

ksundong commented:
고생하셨습니다~
